File: src/test/java/kitchenpos/test/FakeProductRepository.java
package kitchenpos.test;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import kitchenpos.domain.Product;
import kitchenpos.domain.ProductRepository;

public class FakeProductRepository implements ProductRepository {

    private final Map<UUID, Product> storage;

    public FakeProductRepository(Map<UUID, Product> storage) {
        this.storage = storage;
    }

    @Override
    public List<Product> findAllByIdIn(List<UUID> ids) {
        return new ArrayList<>(storage.values().stream()
                .filter(p -> ids.contains(p.getId()))
                .toList());
    }

    @Override
    public Optional<Product> findById(UUID id) {
        return Optional.ofNullable(storage.get(id));
    }

    @Override
    public Product save(Product product) {
        UUID id = UUID.randomUUID();
        product.setId(id);
        storage.put(id, product);
        return product;
    }

    @Override
    public List<Product> findAll() {
        return new ArrayList<>(storage.values());
    }
}

File: src/test/java/kitchenpos/testsupport/FakeProductRepository.kt
package kitchenpos.testsupport

import java.util.Optional
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kitchenpos.domain.Product
import kitchenpos.domain.ProductRepository

class FakeProductRepository : ProductRepository {
    private val storage: MutableMap<UUID, Product> = ConcurrentHashMap()

    override fun findAllByIdIn(list: MutableList<UUID>): MutableList<Product> {
        return storage.values
            .filter { it.id in list }
            .toMutableList()
    }

    override fun findById(productId: UUID): Optional<Product> {
        return Optional.ofNullable(storage[productId])
    }

    override fun findAll(): MutableList<Product> {
        return storage.values.toMutableList()
    }

    override fun save(product: Product): Product {
        storage[product.id] = product
        return product
    }
}